f5_openstack_agent/lbaasv2/drivers/bigip/ftp_profile.py

```diff
@@ -0,0 +1,120 @@
+# coding=utf-8
+# Copyright (c) 2016-2018, F5 Networks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from oslo_log import log as logging
+
+LOG = logging.getLogger(__name__)
+
+
+class FTPProfileHelper(object):
+    """A tool class for all FTP profile process"""
+
+    @staticmethod
+    def enable_ftp(service):
+        listener = service.get('listener')
+        if listener and listener.get('protocol') == 'FTP':
+            return True
+        return False
+
+    # pzhang create ftp
+    @staticmethod
+    def add_profile(service, vip, bigip):
+        ftp_handler = bigip.tm.ltm.profile.ftps.ftp
+        control_port = service.get('listener').get('protocol_port')
+        partition = vip['partition']
+
+        if control_port == 21:
+            LOG.info(
+                "Add default FTP profile: {} for " \
+                "BIGIP: {}".format(
+                   "/Common/ftp", bigip.hostname
+                )
+            )
+            vip['profiles'] = ['/Common/ftp']
+        else:
+            profile_name = FTPProfileHelper.get_profile_name(
+                service)
+            data_port = FTPProfileHelper.get_data_port(
+                control_port)
+            profile = "/" + partition + "/" + profile_name
+
+            profile_exists = ftp_handler.exists(
+                name=profile_name,
+                partition=partition
+            )
+
+            if not profile_exists:
```

Comment:
If it exists, but it content is not what you are expecting, how do you handle that?

Reply:
> If it exists, but it content is not what you are expecting, how do you handle that?

I think it should consider three steps:
  1. figure out how come it has already existed
  2. If any other resources related to this
  3. does it can be manually changed